webapp/src/components/tutorialValidators.tsx

```diff
@@ -0,0 +1,171 @@
+
+const defaultResult: CodeValidationResult = {
+    isValid: true,
+    hint: null,
+};
+
+export function GetValidator(metadata: CodeValidatorMetadata): CodeValidator {
+    switch(metadata.validatorType.toLowerCase()) {
+        case "blocksexistvalidator":
+            return new BlocksExistValidator(metadata.properties);
+        default:
+            console.error(`Unrecognized validator: ${metadata.validatorType}`);
+            return null;
+    }
+}
+
+export function PopulateValidatorCache(metadata: CodeValidationConfig): pxt.Map<CodeValidator>{
+    if(!metadata?.validatorsMetadata) {
+        return null;
+    }
+
+    metadata.validators = {};
+    metadata.validatorsMetadata.forEach(v => {
+        const validator = GetValidator(v);
+        if(validator) {
+            metadata.validators[v.validatorType] = validator;
+        }
+    });
+
+    return metadata.validators;
+}
+
+abstract class CodeValidatorBase implements CodeValidator {
+    enabled: boolean;
+    abstract name: string;
+
+    constructor(properties: {[index: string]: string}) {
+        this.enabled = properties["enabled"]?.toLowerCase() != "false"; // Default to true
+    }
+
+    execute(options: CodeValidationExecuteOptions): Promise<CodeValidationResult> {
+        if (!this.enabled) return Promise.resolve(defaultResult);
+
+        const result = this.executeInternal(options);
+
+        if(!result) {
+            pxt.tickEvent(`codevalidation.detectederror`, {
+                validator: this.name,
+                tutorial: options.tutorialOptions?.tutorial,
+                step: options.tutorialOptions?.tutorialStep,
+            });
+        }
+
+        return result;
+    }
+
+    protected abstract executeInternal(options: CodeValidationExecuteOptions): Promise<CodeValidationResult>;
+}
+
+export class BlocksExistValidator extends CodeValidatorBase {
+    name = "blocksexistvalidator";
+
+    private useHintHighlight = false;
+
+    constructor(properties: { [index: string]: string }) {
+        super(properties);
+        this.useHintHighlight = properties["useanswerkeyhighlight"]?.toLowerCase() === "true";
+    }
+
+    async executeInternal(options: CodeValidationExecuteOptions): Promise<CodeValidationResult> {
+        let missingBlocks: string[] = [];
+        let disabledBlocks: string[] = [];
+        const {parent, tutorialOptions} = options;
+        const stepInfo = tutorialOptions.tutorialStepInfo
+            ? tutorialOptions.tutorialStepInfo[tutorialOptions.tutorialStep]
+            : null;
+        if (!stepInfo) return { isValid: true, hint: "" };
+
+        // Currently, this validator only supports useHintHighlight, so this flag/check is redundant,
+        // but there are plans to support manual specification of blocks in the future.
+        // As such, it felt prudent to still require this flag so tutorial authors don't have to go back and add it later on.
+        if (this.useHintHighlight) {
```

Comment:
`useHintHighlight` does feel like a default case to me -- it might make sense to negate this so you opt out of hintHintlight instead of opt in when this validator is enabled so there's less 'standard' configuration needed?

Reply:
@kiki-lee Thoughts?

Reply:
just to clarify here, what I mean is that instead of having to have

````md
```validation.global
# BlocksExistValidator
* UseAnswerKeyHighlight: true
```
````

would just be

````md
```validation.global
# BlocksExistValidator
```
````

to turn it on, then when we add more features to BlocksExistValidator and you want to set specific blocks would be

````md
```validation.global
# BlocksExistValidator
* DoNotUseAnswerKeyHighlight: true
* CheckFor: spriteSay, myCoolBlock
```
````

Reply:
Makes sense to me, and since there's currently no reason to have this validator with the flag disabled, I'll just remove it altogether for now. We can add one later, once we have other ways to specify blocks to validate.